content/en/blog/low-hanging-performance/index.md

```diff
@@ -0,0 +1,460 @@
+analyzes](https://appliedgo.com/blog/how-to-do-escape-analysis). Use the
+`-gcflags=-m=2` again, but `grep` *escape* lines from the output. That will tell
+you exactly what’s going on with the pointers for every function in the current
+compilation.
+
+Usually, when benchmarking Go code, it’s good to understand what’s going on with
+heap allocations. Just add the following argument e.g., your test benchmark
+compilation, and you get the statistics of allocations in the benchmark run.
+
+```
+go test -benchmem -bench=. <PKG_NAME>
+```
+
+The `-benchmem` flag inserts two columns to benchmarking results:
+
+```
+go test -benchmem -bench=. github.com/lainio/err2/internal/str
+goos: linux
+goarch: arm64
+pkg: github.com/lainio/err2/internal/str
+BenchmarkDecamelRegexp-5          400387              2994 ns/op             666 B/op             19 allocs/op
+BenchmarkDecamel-5               3707656               329.3 ns/op            96 B/op              1 allocs/op
+PASS
+ok      github.com/lainio/err2/internal/str     2.784s
+```
+
+Please note that five (5) columns are now instead of standard three. The extra
+two (rightmost) are the about memory allocations. `B/op` is the average amount of
+bytes per memory allocation in the rightmost column `allocs/op.`
+
+Fewer allocations are better and smaller the size of the allocations. Please
+note that the performance difference between the above benchmark results isn’t
+because of the allocations only. Most of the differences will be explained in
+the following chapters. But still, allocations are something you should be aware
+of, especially about the variable escaping if it leads to heap allocations.
+
+## How Dynamic Are The Inputs?
+
+How much do the variables in your program change, or maybe they are constant?
+Naturally, **the smaller the actual input set of the function, the better chance
+we have to optimize** its performance because the more deterministic the solution
+will be. Also, smaller machine code performs better in modern memory-bound CPUs.
+The same cache rules apply to instructions as variables. CPU doesn’t need to
+access RAM if all the required code is already in the CPU.
+
+The above benchmark results are from two functions that do the same thing. This
+is the regexp version of it (first row in the benchmark results):
+
+```go
+var (
+	uncamel = regexp.MustCompile(`([A-Z]+)`)
+	clean   = regexp.MustCompile(`[^\w]`)
+)
+
+// DecamelRegexp return the given string as space delimeted. Note! it's slow. Use
+// Decamel instead.
+func DecamelRegexp(str string) string {
+	str = clean.ReplaceAllString(str, " ")
+	str = uncamel.ReplaceAllString(str, ` $1`)
+	str = strings.Trim(str, " ")
+	str = strings.ToLower(str)
+	return str
+}
+```
+
+Go’s regexp implementation is known to be relatively slow, but if you think that
+regexp needs its compiler and processor, it’s not so surprising.
+
+The hand-optimized version of the Decamel function is almost ten times faster.
```

Comment:
What exactly are the optimizations? What are the aspects that the reader should pay attention to in the code snippet below?